File: framework/languages/java/samples/java/TicTacToe.SessionGateway/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/session/sessions/handlers/PlayNotificationRelay.java
package systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.handlers;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.samples.tictactoe.sessiongateway.server.play.gamespots.GameNotificationPublisher;
import systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory;

public final class PlayNotificationRelay {
    private PlayNotificationRelay() {
    }

    public static CompletionStage<Void> deliverAsync(String encodedResponse) {
        GameNotificationPublisher.PlayResponse response = GameNotificationPublisher.decode(encodedResponse);
        CompletionStage<Void> tail = CompletableFuture.completedFuture(null);
        for (GameNotificationPublisher.Notification notification : response.notifications()) {
            tail = tail.thenCompose(ignored -> PlayerSessionDirectory
                .find(notification.recipientActorId())
                .map(context -> context.client()
                    .send(notification.payload())
                    .packetName(notification.packetName())
                    .submitAsync())
                .orElseGet(() -> CompletableFuture.completedFuture(null)));
        }
        return tail;
    }

    public static String reply(String encodedResponse) {
        return GameNotificationPublisher.decode(encodedResponse).reply();
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/server/session/sessions/handlers/PlayNotificationRelay.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.handlers

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.play.gamespots.GameNotificationPublisher
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory

object PlayNotificationRelay {
    fun deliverAsync(encodedResponse: String): CompletionStage<Void> {
        val response = GameNotificationPublisher.decode(encodedResponse)
        var tail: CompletionStage<Void> = CompletableFuture.completedFuture(null)
        response.notifications.forEach { notification ->
            tail = tail.thenCompose {
                PlayerSessionDirectory.find(notification.recipientActorId)
                    ?.client()
                    ?.send(notification.payload)
                    ?.packetName(notification.packetName)
                    ?.submitAsync()
                    ?: CompletableFuture.completedFuture(null)
            }
        }
        return tail
    }

    fun reply(encodedResponse: String): String =
        GameNotificationPublisher.decode(encodedResponse).reply
}
